src/sumstats.cpp

```diff
@@ -0,0 +1,185 @@
+// [[Rcpp::depends(RcppArmadillo)]]
+#include <cmath> 
+#include <RcppArmadillo.h>
+#include <Rcpp.h>
+
+// For debugging: Rcpp::Rcout << << std::endl; 
+
+// ----------------------------------------------------------------------------
+
+//' Correlation C++
+//' 
+//' @section Notes:
+//' Verified this function is faster that R's built-in correlation function
+//' for large genotype matrices.
+//'
+//' @param x Numeric matrix.
+//' @return Numeric matrix of correlation among the columns.
+// [[Rcpp::export]]
+SEXP CorCpp(
+  const arma::mat x 
+){
+  return Rcpp::wrap(arma::cor(x));
+};
+
+
+//' Inverse Variance Meta-Analysis
+//' 
+//' @param anno (snps x 1) annotation vector with values in c(0, 1, 2).
+//' @param beta (snps x 1) vector of effect sizes for 
+//'   the coding genetic variants within a gene.
+//' @param ld (snps x snps) matrix of correlations among the genetic variants.
+//' @param se (snps x 1) vector of standard errors for the effect sizes.
+//' @param weights (3 x 1) vector of annotation category weights.
+//' @return Data.frame with the following columns \itemize{
+//' \item anno: gene annotation
+//' \item beta_meta: meta-analyzed effect size
+//' \item se_meta: standard error of the meta-analyzed effect size
+//' }
+// [[Rcpp::export]]
+SEXP IVWCpp(
+  const arma::colvec anno,
+  const arma::colvec beta,
+  const arma::colvec se,
+  const arma::mat ld,
+  const arma::colvec weights
+){
+
+  arma::colvec beta_meta = arma::zeros(3);
+  arma::colvec se_meta = arma::zeros(3);
+  for(int i=0; i<3; i++){
+
+    // Subset LD matrix.
+    arma::uvec key = arma::find(anno == i);
+    arma::mat ld_anno = ld.submat(key, key);
+
+    // Calculate variance.
+    arma::colvec se_anno = se.elem(key);
+    arma::mat d_anno = arma::diagmat(se_anno);
+    arma::mat v_anno = d_anno * ld_anno * d_anno;
+    arma::mat v_anno_inv = arma::inv_sympd(v_anno);
+    double weight_sum = arma::accu(v_anno_inv);
+
+    // Meta-analyze.
+    arma::colvec beta_anno = beta.elem(key);
+    beta_meta(i) = arma::accu(v_anno_inv * beta_anno) / weight_sum / weights(i);
+    se_meta(i) = 1 / sqrt(weight_sum) / weights(i);
+
+  };
+
+  return Rcpp::DataFrame::create(
+    Rcpp::Named("anno")=arma::linspace(0, 2, 3),
+    Rcpp::Named("beta_meta")=beta_meta,
+    Rcpp::Named("se_meta")=se_meta
+  );
+
+};
+
+
+//' Category Correlation
+//' 
```

Comment:
perhaps mention this calculation assumes HWE. I am not sure if HWE is commonly assumed since we are working with rare variants.

Reply:
Thanks for the suggestion, added documentation to this function.